src/main/java/com/scalar/db/storage/cosmos/ResultImpl.java

```diff
@@ -93,69 +92,72 @@ public String toString() {
 
   @VisibleForTesting
   void interpret(Record record, Selection selection, TableMetadata metadata) {
-    record.getPartitionKey().forEach((name, value) -> add(name, value));
-    record.getClusteringKey().forEach((name, value) -> add(name, value));
-
-    // This isn't actual projection...
+    Map<String, Object> recordValues = record.getValues();
     if (selection.getProjections().isEmpty()) {
-      record.getValues().forEach((name, value) -> add(name, value));
+      metadata
+          .getColumns()
+          .forEach(
+              (name, type) -> {
+                add(name, recordValues);
+              });
     } else {
-      Map<String, Object> recordValues = record.getValues();
+      // This isn't actual projection...
       selection
           .getProjections()
           .forEach(
               name -> {
-                if (recordValues.containsKey(name)) {
-                  add(name, recordValues.get(name));
-                }
+                add(name, recordValues);
               });
     }
+
+    metadata.getPartitionKeyNames().forEach(name -> add(name, record.getPartitionKey()));
+    metadata.getClusteringKeyNames().forEach(name -> add(name, record.getClusteringKey()));
   }
 
-  private void add(String name, Object value) {
-    if (metadata.getColumns().containsKey(name)) {
-      values.put(name, convert(value, name, metadata.getColumns().get(name)));
-    } else {
-      throw new InvalidMetadataException("metadata doesn't have the specified column: " + name);
-    }
+  private void add(String name, Map<String, Object> recordValues) {
+    values.put(name, convert(recordValues.get(name), name, metadata.getColumns().get(name)));
   }
 
   private Optional<Key> getKey(Set<String> names) {
     List<Value> list = new ArrayList<>();
     for (String name : names) {
       Value value = values.get(name);
-      if (value == null) {
-        LOGGER.warn("full key doesn't seem to be projected into the result");
-        return Optional.empty();
-      }
```

Comment:
This case doesn't happen anymore because the `values` always have the corresponding entry for the`name`.
All keys' entries are put in `interpret()`.